src/box/iproto.cc

```diff
@@ -1851,7 +2226,88 @@ net_send_msg(struct cmsg *m)
+	 * `net_finish_disconnect`
+	 */
+	if (stream_id == 0 || con->state == IPROTO_CONNECTION_CLOSED)
+		goto send_msg;
+
+	pos = mh_streams_find(con->streams, stream_id, 0);
+	assert(pos != mh_end(con->streams));
+	stream = *mh_streams_node(con->streams, pos);
+
+	if (rlist_empty(&stream->pending_list)) {
+		/*
+		 * If no more messages for the current stream
+		 * and no transaction started, then delete it.
+		 */
+		if (stream->txn == NULL) {
+			mh_streams_remove(con->streams, (const struct stream **)
+					  &stream, 0);
+			iproto_stream_delete(stream);
+		}
+	} else {
+		/*
+		 * If there are new messages for this stream
+		 * then schedule their processing.
+		 */
+		struct iproto_msg * next =
+			rlist_first_entry(&stream->pending_list,
+					  struct iproto_msg,
+					  in_pending_list);
+		assert(next != NULL);
+		/*
+		 * It's very important to update wpos before push message
+		 * to tx thread. Otherwise, there may be a situation where
+		 * we reset obuf, but not resetting the corresponding svp.
+		 * For example we have two streams each of which processes
+		 * transaction. And there are several messages in each stream
+		 * which are waited in `pending_list`. For example we have the
+		 * following sequence of requests:
+		 * (0) Some previous requests which was ended long ago.
+		 * (1) Begin in stream_1
+		 * (2) Begin in stream_2
+		 * (3) Some requests in stream_1 and stream_2
+		 * (4) Commit in stream_1
+		 * (5) Commit in stream 2
+		 * (6) Request in stream_1
+		 * All previous requests came at the same time and have the
+		 * same wpos. When (1) starts to be processed in tx thread, he
+		 * reset obuf, which was used in (0). Svp of this obuf was
+		 * previously reseted in `iproto_flush`. All this requests will
+		 * use the same one obuf.
+		 * (7) Request in stream_2
+		 * Request (7) comes later, after one of the previous requests
+		 * is processed, so it can have (and has for our example) other
+		 * wpos. If transaction in stream_1 conflicts with transaction
+		 * in stream_2, (4) finishes later then (5). So request (7)
+		 * begins before request (6), and because it has other wpos it
+		 * change tx.p_obuf in connection and reset obuf, which was
+		 * used by requests (1 -5). So far, so good, since query (7)
+		 * has the actual wpos. But if (6) starts immidiatly, and we
+		 * don't update it wpos in next line, it will have incorrect
+		 * wpos. It will update tx.p_obuf again and reset obuf which
+		 * is used by request (7), moreover svp of (1-5) was not
+		 * reseted, but (6) used obuf to which this not reseted svp
+		 * belongs!. If we update wpos here (6) uses same obuf with (7),
+		 * svp of obuf, which was used in (1 - 5) will be reseted in
+		 * `iproto_flush` after (7) and (6) are processed (expected
+		 * behavior).
+		 */
+		next->wpos = con->wpos;
+		cpipe_push_input(&con->iproto_thread->tx_pipe,
+				 &next->base);
+		cpipe_flush_input(&con->iproto_thread->tx_pipe);
```

Comment:
As for me all above still looks like a bug. I don't believe we have to do such thing. Maybe Vlad or Vova will accept this, OK then.

Reply:
That's was a huge comment by I still don't understand what that svp is for.